### ISSUE TITLE
multi-monitor setup broken (X11 issue?)

### ISSUE THREAD

**Issue by dvzrv** [qtile/qtile#1041]
For some time now, I'm unable to setup a multi-monitor setup, or use an external monitor at all for my laptop.
It reproducably leads to visual glitches and being unable to use qtile - although it is still responsive and able to restart, it is not showing (or rather not drawn) correctly. It shows up as tiles made up of parts of the screens.

I'm on Arch Linux with qtile 0.10.7, xorg-server 1.19.3, python 3.6.

So far I've used xrandr, lxrandr, arandr to define the setup (laptop with one external screen, HDMI or VGA, or external monitor only), but they all lead to the same visual glitches:

external and internal with same resolution, side by side:
![qtile-xserver-foo2](https://cloud.githubusercontent.com/assets/432519/24952331/67061c56-1f77-11e7-9fda-82500201a5a9.png)

external with full resolution side by side with internal:
![qtile-xserver-foo-higher-solution2](https://cloud.githubusercontent.com/assets/432519/24952333/6709e804-1f77-11e7-868b-166db41617b9.png)

only external:
![qtile-xserver-foo-only-external](https://cloud.githubusercontent.com/assets/432519/24952332/670657e8-1f77-11e7-8a42-7b2a2e45379b.png)

The size of the internal monitor is drawn in a rectangle to the top-left side of the screen, mostly qtile is not properly drawn within that though, and the rest of the screen is unusable with a frozen still-image.

This issue also applies to awesome wm, but is not happening on xfce!

**Comment by frostidaho** [comment]
What do qtile's screen objects look like?
```shell
echo screens | qshell
```

And how does the geometry of these screen objects compare to the output from `xrandr`?

**Comment by dvzrv** [comment]
@frostidaho from what I understand, they seem to be just fine:

internal setup:

```
> [{'gaps': {'bottom': None,
           'left': None,
           'right': None,
           'top': (0, 0, 1366, 20)},
  'group': 'shell',
  'height': 768,
  'index': 0,
  'width': 1366,
  'x': 0,
  'y': 0}]
> 
```

```
Screen 0: minimum 8 x 8, current 1366 x 768, maximum 32767 x 32767
LVDS1 connected primary 1366x768+0+0 (normal left inverted right x axis y axis) 280mm x 160mm
   1366x768      60.00*+
   1280x720      60.00  
   1024x768      60.00  
   1024x576      60.00  
   960x540       60.00  
   800x600       60.32    56.25  
   864x486       60.00  
   640x480       59.94  
   720x405       60.00  
   680x384       60.00  
   640x360       60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP3 disconnected (normal left inverted right x axis y axis)
HDMI1 connected (normal left inverted right x axis y axis)
   1920x1200     59.95 +
   1600x1200     60.00  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   640x480       75.00    72.81    66.67    59.94  
   720x400       70.08  
HDMI2 disconnected (normal left inverted right x axis y axis)
HDMI3 disconnected (normal left inverted right x axis y axis)
VGA1 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

internal and exernal on full resolution, side by side:

```
> [{'gaps': {'bottom': None,
           'left': None,
           'right': None,
           'top': (1920, 0, 1366, 20)},
  'group': 'web',
  'height': 768,
  'index': 0,
  'width': 1366,
  'x': 1920,
  'y': 0},
 {'gaps': {'bottom': None,
           'left': None,
           'right': None,
           'top': (0, 0, 1920, 20)},
  'group': 'shell',
  'height': 1200,
  'index': 1,
  'width': 1920,
  'x': 0,
  'y': 0}]
> 
```

```
Screen 0: minimum 8 x 8, current 3286 x 1200, maximum 32767 x 32767
LVDS1 connected primary 1366x768+1920+0 (normal left inverted right x axis y axis) 280mm x 160mm
   1366x768      60.00*+
   1280x720      60.00  
   1024x768      60.00  
   1024x576      60.00  
   960x540       60.00  
   800x600       60.32    56.25  
   864x486       60.00  
   640x480       59.94  
   720x405       60.00  
   680x384       60.00  
   640x360       60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP3 disconnected (normal left inverted right x axis y axis)
HDMI1 connected 1920x1200+0+0 (normal left inverted right x axis y axis) 520mm x 320mm
   1920x1200     59.95*+
   1600x1200     60.00  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   640x480       75.00    72.81    66.67    59.94  
   720x400       70.08  
HDMI2 disconnected (normal left inverted right x axis y axis)
HDMI3 disconnected (normal left inverted right x axis y axis)
VGA1 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

only external screen on full resolution:

```
> [{'gaps': {'bottom': None,
           'left': None,
           'right': None,
           'top': (0, 0, 1920, 20)},
  'group': 'shell',
  'height': 1200,
  'index': 0,
  'width': 1920,
  'x': 0,
  'y': 0}]
> 
```

```
Screen 0: minimum 8 x 8, current 1920 x 1200, maximum 32767 x 32767
LVDS1 connected (normal left inverted right x axis y axis)
   1366x768      60.00 +
   1280x720      60.00  
   1024x768      60.00  
   1024x576      60.00  
   960x540       60.00  
   800x600       60.32    56.25  
   864x486       60.00  
   640x480       59.94  
   720x405       60.00  
   680x384       60.00  
   640x360       60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP3 disconnected (normal left inverted right x axis y axis)
HDMI1 connected primary 1920x1200+0+0 (normal left inverted right x axis y axis) 520mm x 320mm
   1920x1200     59.95*+
   1600x1200     60.00  
   1280x1024     75.02    60.02  
   1280x960      60.00  
   1152x864      75.00  
   1024x768      75.03    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   640x480       75.00    72.81    66.67    59.94  
   720x400       70.08  
HDMI2 disconnected (normal left inverted right x axis y axis)
HDMI3 disconnected (normal left inverted right x axis y axis)
VGA1 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)

```

**Comment by frostidaho** [comment]
Hmm so your laptop is 1366x768 and your other monitor is 1920x1200, and you have a 20pixel bar on the top right? If that's the case everything appears to be at least consistent between qtile and the xserver. I can't say I have any idea as to what the issue might be, but I'd be willing to bet the problem doesn't lie with qtile. After all you mentioned your problem also occurs with awesomewm

**Comment by dvzrv** [comment]
@frostidaho correct.
I've now also checked the whole setup with and without loading of intel-ucode, using and not using of [autorandr](https://github.com/phillipberndt/autorandr), but it doesn't make a difference.

Maybe worth noting: I'm using lightdm (1.22.0) to start my sessions.

**Comment by dvzrv** [comment]
Okay... after some more debugging, I finally found out, what the problem was.

For some unreproducable reason awesome and qtile will give the above results, when the Intel driver **i915** has `sna` acceleration method (now the implicit default!) set.

`
Option      "AccelMethod"  "sna"
`

Setting the acceleration method to `uxa` fixes this:

`Option      "AccelMethod"  "uxa" `

As mentioned above, this does not happen under xfce... which is very strange, given the fact, that it is a graphics driver option, changing this behavior!

I have to note though, that gdm and lightdm also behave differently, when using `sna` or `uxa`.
When on `sna`, the desktop managers clone the output (mouse shown on both screens). The login window (with username) is however only duplicated using lightdm in that case.
When on `uxa` the desktop managers extend to the external screen (mouse only shown once). The login window is also only shown once then (for both DMs).

I know this is most likely a hardware + driver issue, but what I can't get into my head is, that xfce works well...
What do they do differently there?

**Comment by flacjacket** [comment]
Hmm, that's interesting. I have intel graphics too, but haven't run into this issue. I thought I used to use SNA when I had the DDX driver, but I haven't used it in a while now. Are you using the DDX driver? You could try removing it, tho I guess you have found a workaround.

**Comment by dvzrv** [comment]
@flacjacket I think the "uxa" AccelMethod is just a dirty workaround, as it means that a lot of other things break with it. Restarting just to use an external monitor is also no way to go :(
It seems that upstream has a lot of bug reports on broken functionality with the newer version of the driver, especially `sna` not properly dealing with multi-monitor setups.

If I restart the session (i.e. the DM), it works with `uxa` for the resolutions still supported by my graphics card. If monitors with higher default resolutions are connected, this breaks again (in the same way it breaks for `sna`.

I can only hope, that Intel will get their act together on this...

**Comment by dvzrv** [comment]
I have found the culprit!

It was due to a faulty configuration file for [compton](https://github.com/chjj/compton), that X11 somehow messed up the screen setup.
Somehow starting compton with my configuration file leads to this crash (but seemingly also only sometimes). I disabled compton and had no issues. Even starting compton with

`/usr/bin/compton -d "$DISPLAY" -CGb &`

doesn't produce this error, so "standard" settings should be fine completely.

Granpa X11...
